### DISCUSSION TITLE
Mapping Solutions and Bounds from Original Problem Space to Preprocessed and Back

### DISCUSSION THREAD

**Discussion by spkelle2** [coin-or/Cgl#88] [Q&A] [answered]
I saw a few lines of code starting [here](https://github.com/coin-or/Cbc/blob/master/src/CbcModel.cpp#L2053) within Cbc that appear to map a solution from the original problem space to the preprocessed problem space. Am I correct? Is this transformed solution valid for the preprocessed problem?

While we're on this topic, does `CglPreprocess::presolve()` only drop variables, or does it create some as well? 

Further, how can I get the dual bound of an instance if I called branchAndBound on its preprocessed counterpart? It looks like I can do [this](https://github.com/coin-or/Cbc/blob/master/src/CbcSolver.cpp#L9113) (defined [here](https://github.com/coin-or/Cbc/blob/master/src/CbcModel.cpp#L7723)), but is that valid? Are we guaranteed to have the same dual bound in the original instance as we found in the preprocessed instance?

**Comment by jjhforrest** [comment, marked as answer]
The transformed solution is valid for preprocessed problem - however it is more flexible not to use "strategy" but to use CglPreprocess as in CbcSolver.cpp.  
I think you mean CglPreprocess::preProcess - the presolve function returns a model saved during preProcess.  preProcess can create variables - it rarely does.
The dual bound is valid.

**Reply by spkelle2** [reply]
You're right that I mean CglPreprocess::preProcess. Thank you for that call out. This saves some unit testing, so thank you for getting back to me on what I can expect with the preprocessor!
